core/src/main/java/org/jboss/pnc/rex/core/GenericVertxHttpClient.java

```diff
@@ -148,24 +154,37 @@ private Uni<HttpResponse<Buffer>> handleRequest(Uni<HttpResponse<Buffer>> uni, C
                 );
 
         // cases when http request method itself fails (unreachable host)
-        uni = uni.onFailure(this::skipOn425)
+        uniWithCtx = uniWithCtx.onFailure(this::skipOnResponse)
                 .invoke(t ->  {
                     if (t instanceof RequestRetryException) {
                         log.warn("HTTP-CLIENT : Http call failed. RETRYING.");
-                    } else {
-                        log.warn("HTTP-CLIENT : Http call failed. RETRYING. Reason: ", t);
+                    } else if (t instanceof HttpResponseException hre) {
+                        log.warn("HTTP-CLIENT : Http call failed ({}). RETRYING.", hre.getStatusCode());
                     }
                 });
-
-        uni = requestRetryPolicy
-                .applyToleranceOn(this::skipOn425, uni);
+        uniWithCtx = requestRetryPolicy.applyToleranceOn(this::skipOnResponse, uniWithCtx);
 
         // fallback to connection unreachable after an exception
-        uni = uni.onFailure().invoke(onConnectionUnreachable)
+        BiFunction<Throwable, Context, Void> onError = (t, ctx) -> {
+            if (!(t instanceof HttpResponseException) && t.getCause() instanceof HttpResponseException) {
+                // prevent java.lang.IllegalStateException in case of exhausted retries due to expire-in limit
+                onConnectionUnreachable.accept(t.getCause());
+            } else {
+                onConnectionUnreachable.accept(t);
+            }
+            HttpResponse<Buffer> response = ctx.get(CTX_RESPONSE_KEY);
+            log.warn("Request failed with code: {}, body: {}.", response.statusCode(), response.bodyAsString());
```

Comment:
I'd log **before** calling `onConnectionUnreachable()`. The reason is that onConnectionUnreachable is usually contains a call to TaskController which will print a lot of logs and may trigger transitions for other Tasks. So this log line will lose context.

Reply:
Good catch! Right, there will be no response as one does not exists. 
I would still leave it here, to not make the Uni chaining to complicated. I rather added a check if there is a request in the context. WDYT?